lib/tracecontext/tracestate.js

```diff
@@ -0,0 +1,346 @@
+/**
+ * Class for Managing Tracestate
+ *
+ * Class that creates objects for managing trace state.
+ * This class is capable of parsing both tracestate strings
+ * and tracestate binary representations, allowing clients
+ * to get and set values in a single list-member/namespace
+ * while preserving values in the other namespaces.
+ *
+ * Capable of working with either the binary of string
+ * formatted tracestate values.
+ *
+ * Usage:
+ *   const tracestate = TraceState.fromStringFormatString(headerTracestate, 'es')
+ *   tracestate.setValue('s',1)
+ *   const newHeader = tracestate.toW3cString()
+ */
+class TraceState {
+  constructor (sourceBuffer, listMemberNamespace = 'es', defaultValues = {}, logger) {
+    if (!this._validateVendorKey(listMemberNamespace)) {
+      throw new Error('Vendor namespace failed validation.')
+    }
+
+    // buffer representation of the trace state string.
+    // The initial value of this.buffer will keep the
+    // values set in the listMemberNamespace list-member,
+    // but as soon as an initial value is set (via setValue)
+    // then the listMemberNamespace values will be removed
+    // from this.buffer and stored in the this.values.  While
+    // slightly more complicated, this allows us to maintain
+    // the order of list-member keys in an un-mutated tracestate
+    // string, per the W3C spec
+    this.buffer = sourceBuffer
+
+    this.listMemberNamespace = listMemberNamespace
+
+    // values for our namespace, set via setValue to
+    // ensure names conform
+    this.values = {}
+    for (const [key, value] of Object.entries(defaultValues)) {
+      this.setValue(key, value)
+    }
+
+    // if no logger, then fallback to agent singleton
+    logger = logger || require('../..').logger
+    this.logger = logger
+  }
+
+  setValue (key, value) {
+    const strKey = String(key)
+    const strValue = String(value)
+    if (!this._validateElasicKeyAndValue(strKey, strValue)) {
+      this.logger.trace('could not set tracestate key, invaliad characters detected')
+      return false
+    }
+    const isFirstSet = (Object.keys(this.values).length === 0)
+    const oldValue = this.values[strKey]
+    this.values[strKey] = value
+
+    // per: https://github.com/elastic/apm/blob/d5b2c87326548befcfec6731713932a00e430b99/specs/agents/tracing-distributed-tracing.md
+    // If adding another key/value pair to the es entry would exceed the limit
+    // of 256 chars, that key/value pair MUST be ignored by agents.
+    // The key/value and entry separators : and ; have to be considered as well.
+    const serializedValue = this._serializeValues(this.values)
+    if (serializedValue.length > 256 && (typeof oldValue === 'undefined')) {
```

Comment:
May be merge the below two if statements?

Reply:
_nod_ that seems like an OK change.  I generally prefer stand alone `if` statements (no `else/else if` etc.) to avoid super complicated conditional logic -- but this isn't that complicated _and_ with the `else` we avoid a few extra instructions on ever `set`.  Made this change.